File: src/common/py/integration/authorization/strategies/authorization_strategies.py
import typing

from .authorization_strategies_catalog import AuthorizationStrategiesCatalog
from .authorization_strategy import AuthorizationStrategy
from .oauth2 import (
    OAuth2Strategy,
    create_oauth2_strategy,
)
from ....component import BackendComponent
from ....core import logging
from ....data.entities.authorization import AuthorizationToken
from ....data.entities.user import UserToken
from ....services import Service


def register_authorization_strategies() -> None:
    """
    Registers all available authorization strategies.

    When adding a new strategy, always register it here.
    """

    # New strategies go here
    AuthorizationStrategiesCatalog.register_item(
        OAuth2Strategy.Strategy, create_oauth2_strategy
    )

    # Print all available strategies for debugging purposes
    names: typing.List[str] = []
    for name, _ in AuthorizationStrategiesCatalog.items():
        names.append(name)
    logging.debug(f"Registered authorization strategies: {'; '.join(names)}")


def create_authorization_strategy(
    comp: BackendComponent,
    svc: Service,
    strategy: str,
    config: typing.Any,
    *,
    user_token: UserToken | None = None,
    auth_token: AuthorizationToken | None = None,
) -> AuthorizationStrategy:
    """
    Creates an authorization strategy using the specified identifier.

    Args:
        comp: The global component.
        svc: The service to use for message sending.
        strategy: The strategy identifier.
        config: The strategy configuration as an arbitrary record.
        user_token: An optional user token.
        auth_token: An optional authorization token.

    Returns:
        The newly created strategy.
    """
    if strategy == "":
        raise RuntimeError("No authorization strategy has been provided")

    strategy_creator = AuthorizationStrategiesCatalog.find_item(strategy)
    if strategy_creator is None:
        raise RuntimeError(f"The authorization strategy '{strategy}' couldn't be found")

    return strategy_creator(
        comp,
        svc,
        config,
        user_token=user_token,
        auth_token=auth_token,
    )

File: src/common/web/integration/authorization/strategies/AuthorizationStrategies.ts
import { WebComponent } from "../../../component/WebComponent";
import logging from "../../../core/logging/Logging";
import { Service } from "../../../services/Service";
import { AuthorizationStrategiesCatalog } from "./AuthorizationStrategiesCatalog";
import { AuthorizationStrategy } from "./AuthorizationStrategy";
import { createOAuth2Strategy, OAuth2Strategy } from "./oauth2/OAuth2Strategy";

/**
 * Registers all available authorization strategies.
 *
 * When adding a new strategy, always register it here.
 */
export function registerAuthorizationStrategies(): void {
    // New strategies go here
    AuthorizationStrategiesCatalog.registerItem(OAuth2Strategy.Strategy, createOAuth2Strategy);

    // Print all available strategies for debugging purposes
    const names = Object.keys(AuthorizationStrategiesCatalog.items);
    logging.debug(`Registered authorization strategies: ${names.join("; ")}`);
}

/**
 * Creates an authorization strategy using the specified identifier.
 *
 * @param comp - The global component.
 * @param svc - The service to use for message sending.
 * @param strategy - The strategy identifier.
 * @param config - The host strategy configuration as an arbitrary record.
 *
 * @returns - The newly created strategy.
 */
export function createAuthorizationStrategy(comp: WebComponent, svc: Service, strategy: string, config: Record<string, any>): AuthorizationStrategy {
    if (!strategy) {
        throw new Error("No authorization strategy has been provided");
    }

    const strategyCreator = AuthorizationStrategiesCatalog.findItem(strategy);
    if (!strategyCreator) {
        throw new Error(`The authorization strategy '${strategy}' couldn't be found`);
    }

    return strategyCreator(comp, svc, config);
}
